File: algorithms/cpp/H-Index II.cpp
class Solution {
public:
    int hIndex(vector<int>& citations) {
        int n = citations.size();
        
        int l = 0, h = n - 1;
        while(l <= h) {
            int m = (l + h) / 2;
            if(n - m > citations[m]) {
                l = m + 1;
            } else if(n - m < citations[m]) {
                h = m - 1;
            }
            else {
                return n - m;
            }
        }
        
        return n - l;
    }
};

File: algorithms/java/H-Index II.java
public class Solution {
    public int hIndex(int[] citations) {
        int index = 0;
        for(int i = citations.length - 1;i >= 0;--i) {
            if(citations[i] >= index + 1) {
                index++;
                continue;
            } else {
                return index;
            }
        }
        
        return citations.length;
    }
}

public class Solution {
    public int hIndex(int[] citations) {
        int l = 0, r = citations.length;
        while(l < r) {
            int m = l + (r - l + 1) / 2;
            if(citations[citations.length - m] >= m) {
                l = m;
            } else {
                r = m - 1;
            }
        }
        
        return l;
    }
}